onnxruntime/core/optimizer/constant_sharing.cc

```diff
@@ -0,0 +1,222 @@
+  return initializer_name.rfind(SHARED_INITIALIZER_PREFIX, 0) == 0;
+}
+
+bool PrepareInputPortsToReplace(Graph& graph, const NodeArg* origin_initializer_node_arg,
+                                InlinedHashMap<const Node*, InlinedVector<int>>& consumer_node_to_input_ports_map) {
+  std::vector<const Node*> consumers = graph.GetConsumerNodes(origin_initializer_node_arg->Name());
+
+  // If usage is from subgraph, skip it now, can be extended to support if there is a need.
+  bool found_subgraph_usage = false;
+  for (const Node* const_node : consumers) {
+    for (int i = 0; i < static_cast<int>(const_node->ImplicitInputDefs().size()); ++i) {
+      if (const_node->ImplicitInputDefs()[i] == origin_initializer_node_arg) {
+        found_subgraph_usage = true;
+        break;
+      }
+    }
+
+    if (found_subgraph_usage) {
+      break;
+    }
+
+    // Iterate all input defs to replace those that are equal to origin_initializer_node_arg,
+    // Then it would be safe to remove the consumer node aferwards.
+    for (int i = 0; i < static_cast<int>(const_node->InputDefs().size()); ++i) {
+      if (const_node->InputDefs()[i] == origin_initializer_node_arg) {
+        consumer_node_to_input_ports_map[const_node].push_back(i);
+      }
+    }
+  }
+
+  return found_subgraph_usage;
+}
+
+// Replace all consumer nodes to use shared initializers.
+void ReplaceInputsToUseSharedInitializer(Graph& graph,
+                                         const InlinedHashMap<const Node*, InlinedVector<int>>&
+                                             consumer_node_to_input_ports_map,
+                                         const NodeArg* origin_initializer_node_arg,
+                                         NodeArg* shared_initializer_node_arg) {
+  for (auto it = consumer_node_to_input_ports_map.begin(), end = consumer_node_to_input_ports_map.end();
+       it != end; ++it) {
+    Node* node = graph.GetNode(it->first->Index());
+    // Iterate all input defs to replace those that are equal to origin_initializer_node_arg,
+    // Then it would be safe to remove the consumer node.
+    for (int input_index : it->second) {
+      graph_utils::ReplaceNodeInput(*node, input_index, *shared_initializer_node_arg);
+    }
+    graph.RemoveConsumerNode(origin_initializer_node_arg->Name(), node);
+
+    //  Add consumer ref count for shared scalar initializer.
+    std::vector<const Node*> consumers = graph.GetConsumerNodes(shared_initializer_node_arg->Name());
+    if (std::find(consumers.begin(), consumers.end(), node) == consumers.end()) {
+      graph.AddConsumerNode(shared_initializer_node_arg->Name(), node);
+    }
+  }
+
+  // Remove the initializer if no other consumer nodes.
+  if (graph.GetConsumerNodes(origin_initializer_node_arg->Name()).size() == 0) {
+    graph.RemoveInitializedTensor(origin_initializer_node_arg->Name());
+  }
+}
+
+#define DispatchOnONNXTensorProtoDataTypeWithReturn(data_type, retval, function, ...) \
```

Comment:
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es31-dont-use-macros-for-constants-or-functions

We have type-based dispatchers that can be used instead. See usage of utils::MLTypeCallDispatcherFromTypeList. Create a templatized function or functor that does the work and plugin a boost::mp11::mp_list of supported types.